### PR TITLE
Add upper bounds report to dashboard

### LINKED ISSUE #89: Implement upper bounds check for dashboard

### PR DESCRIPTION
**Pull request by elharo**
@garrettjonesgoogle @netdpb fix #89

### CODE REVIEW

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 1]
```diff
@@ -104,17 +104,40 @@ private static void generateReport(Configuration configuration, Path output, Art
...
       DependencyGraph graph =
           DependencyGraphBuilder.getCompleteDependencies(artifact);
-      List<Update> updates = graph.findUpdates();
-   
+      List<Update> updates = graph.findUpdates();      
+      
+      // picks versions according to Maven rules
+      DependencyGraph transitiveDependencies = DependencyGraphBuilder.getTransitiveDependencies(artifact);
+      
+      Map<String, String> expectedVersionMap = graph.getHighestVersionMap();
```
[the remark this answers is not in the document]
Comment (elharo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: done

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraph.java [hunk 2]
```diff
@@ -142,5 +139,19 @@ void addPath(DependencyPath path) {
     
     return new ArrayList<>(upgrades);
   }
+
+  public Map<String, String> getHighestVersionMap() {
+    Map<String, Collection<String>> input = versions.asMap();
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraph.java]: I think I would find that harder to read.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraph.java [hunk 3]
```diff
@@ -142,5 +139,19 @@ void addPath(DependencyPath path) {
...
+    Map<String, String> output = new HashMap<>();
+    
+    VersionComparator comparator = new VersionComparator();
+    
+    for (Map.Entry<String, Collection<String>> entry : input.entrySet()) {
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraph.java]: I might also find that harder to read and definitely harder to debug since there's nowhere to put breakpoints.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  @Test
+  // beam has a more complex dependency graph that hits some corner cases
+  public void testFindUpdates_beam()
+      throws DependencyCollectionException, DependencyResolutionException {    
+
+    DefaultArtifact beam =
+        new DefaultArtifact("org.apache.beam:beam-sdks-java-io-google-cloud-platform:2.5.0");
+    DependencyGraph graph = DependencyGraphBuilder.getCompleteDependencies(beam);
+    
+    graph.findUpdates();
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java]: Yep, nothing is thrown. This test is here because it used to fail. Only moved in this PR. In particular there are some weirdnesses in the Beam dependency graph where resolution of the correct artifact requires detecting which operating system you're running on.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  // a non-Google dependency graph that's well understood and thus useful for debugging
+  public void testJaxen()
+      throws DependencyCollectionException, DependencyResolutionException {    
+
+    DefaultArtifact jaxen =
+        new DefaultArtifact("jaxen:jaxen:1.1.6");
+    DependencyGraph graph = DependencyGraphBuilder.getCompleteDependencies(jaxen);
+    
+    List<Update> updates = graph.findUpdates();
+    Assert.assertEquals(5, updates.size());
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java]: done

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    
+    Map<String, String> complete = completeDependencies.getHighestVersionMap();    
+    Map<String, String> transitive =
+        transitiveDependencies.getHighestVersionMap();
+    Set<String> completeKeyset = complete.keySet();
+    Set<String> transitiveKeySet = transitive.keySet();
+    
+    // The complete dependencies sees a path to com.google.j2objc:j2objc-annotations that's
+    // been removed in newer versions.
+    Assert.assertTrue(completeKeyset.contains("com.google.j2objc:j2objc-annotations"));
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java]: done

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphTest.java [hunk 7]
```diff
@@ -58,12 +62,25 @@ public void setUp() {
...
     graph.addPath(path3);
     graph.addPath(path4);
     graph.addPath(path5);
+    graph.addPath(path6);
+  }
+  
+  @Test
+  public void testGetHighestVersionMap() {
+    Map<String, String> map = graph.getHighestVersionMap();
+    Assert.assertEquals("2", map.get("com.google:baz"));
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphTest.java]: Again, this reads more straightforwardly to me. assertThat(map).containsExactly(...) seems particularly challenging to follow since it doesn't distinguish between keys and values except by position.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraph.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (netdpb) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraph.java]: But you're recreating the comparator here. The point of using a `TreeMultimap` is to get that sorted-set behavior out of the collections.

Up to you, I guess.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (netdpb) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java]: Can you add a comment to this effect here?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraph.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (netdpb) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraph.java]: But there's less (nothing?) to debug, because the `TreeMultimap` has already found the max for you.

Up to you.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java [hunk 8]
```diff
@@ -120,8 +120,8 @@ public void testGrpcAuth()
     Set<String> transitiveKeySet = transitive.keySet();
     
     // The complete dependencies sees a path to com.google.j2objc:j2objc-annotations that's
-    // been removed in newer versions.
-    Assert.assertTrue(completeKeyset.contains("com.google.j2objc:j2objc-annotations"));
+    // been removed in newer versions so this is not bidirectional set equality.
+    Assert.assertTrue(complete.containsKey("com.google.j2objc:j2objc-annotations"));
```
Comment (netdpb) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java]: `assertThat(complete).containsKey(...)`

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java]: done

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java [hunk 9]
```diff
@@ -74,8 +74,12 @@ public void testFindUpdates()
         + "upgrade com.google.http-client:google-http-client:1.19.0 to 1.23.0");
   }
   
+  // Beam has a more complex dependency graph that hits some corner cases.
+  // In particular it pulls in Netty, which pulls in native code, the 
+  // exact artifact depending on which operating system you're running on.
+  // This is the static block that sets the os.detected.classifier system property.
```
Comment (netdpb) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java]: Which is the static block that "this" refers to? Take *what* out and this test will fail?

Can you add a comment, maybe on line 90, like `// should not throw`?

Comment (elharo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java]: done
